Starting the document on TLA+ idioms (#325)
informalsystems/apalache: docs/idiomatic/README.md

# Idiomatic TLA+

> **Authors:** Igor Konnov + (who likes to contribute?)

Author: ```suggestion
**Authors:** Shon Feder, Igor Konnov + (who likes to contribute?)
```

# Idiomatic TLA+

**Authors:** Igor Konnov + (who likes to contribute?)

_This document is under construction.
If you like to contribute, open a pull request._

## Introduction

In this document, we collect specification idioms that aid us in writing TLA+
specifications that are:

1. understood by distributed system engineers,
1. understood by verification engineers, and
1. understood by automatic analysis tools such as the Apalache model checker.

If you believe, that the above points are contradictory when put together, it is
to some extent true. TLA+ is an extremely general specification language. As a

> result, it is easy to write a short specification that leaves puzzled a human
> reader. It is even easier to write a (syntactically correct) specification that

Reviewer: ```suggestion
result, it is easy to write a short specification that leaves a human reader
puzzled . It is even easier to write a (syntactically correct) specification that
```

> turns to dust any program reasoning about TLA+. 

Reviewer: ```suggestion
turns any program trying to reason about TLA+ to dust. 
```

Nevertheless, we find TLA+ quite useful when writing concise specifications of
distributed protocols at [Informal Systems]. Other specification languages --
especially, those designed for software verification -- would require us to
introduce unnecessary book-keeping details that would both obfuscate the
protocols and make their verification harder. However, we do not always need

> _"all power of mathematics"_, so we find it useful to introduce additional

Reviewer: ```suggestion
_"all the power of mathematics"_, so we find it useful to introduce additional
```

structure in TLA+ specifications.

Below, we summarize the idioms that help us in maintaining that structure.  As
a bonus, these idioms usually aid the Apalache model checker in analyzing the
specifications. Our idioms are quite likely different from the original ideas
of [Leslie Lamport] (the author of TLA+).
So it is useful to read Lamport's [Specifying Systems]. Importantly, these are

> _idioms_, not the rules set in stone. If you believe that one of those idioms

Reviewer: ```suggestion
_idioms_, not rules set in stone. If you believe that one of those idioms
```

does not work for you in your specific setting, don't follow it.

If this is the first page where you encounter the word "TLA+", we do not

> recommend continue reading the material. It is better to start with [The TLA+

Reviewer: ```suggestion
recommend that you continue to read the material. It is better to start with [The TLA+
```

Video Course](http://lamport.azurewebsites.net/video/videos.html) by Leslie
Lamport.  Once you have understood the basics and tried the language, it makes
sense to ask the question: "How do I write a specification that other people

> understand?".  We believe that many TLA+ users reinvent the rules that are
> similar to our idioms.  By providing you with a bit of thought, we hope to

Reviewer: ```suggestion
understand?".  We believe that many TLA+ users reinvent rules that are
similar to our idioms.  By providing you with a bit of guidance, we hope to
```

> reduce your discomfort when learning TLA+.

Reviewer: ```suggestion
reduce your discomfort when learning more advanced TLA+.
```
